File: python/tree.py
class TreeNode:
    def __init__(self, label):
        self.label = label
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.root = None

    def set_root(self, node):
        self.root = node

    def get_root(self):
        return self.root

    def set_left(self, parent, child):
        if parent:
            parent.left = child

    def set_right(self, parent, child):
        if parent:
            parent.right = child

    def preorder(self, node, visit):
        if node:
            visit(node.label)
            self.preorder(node.left, visit)
            self.preorder(node.right, visit)

    def inorder(self, node, visit):
        if node:
            self.inorder(node.left, visit)
            visit(node.label)
            self.inorder(node.right, visit)

    def postorder(self, node, visit):
        if node:
            self.postorder(node.left, visit)
            self.postorder(node.right, visit)
            visit(node.label)

    def is_empty(self):
        return self.root is None

    def clear(self):
        self.root = None

File: src/tree.cpp
#include "tree.hpp"

TreeNode::TreeNode(char lbl) : label(lbl), left(nullptr), right(nullptr) {}

TreeNode::~TreeNode() {
    delete left;
    delete right;
}

Tree::Tree() : root(nullptr) {}

Tree::~Tree() {
    clear();
}

void Tree::setRoot(TreeNode* node) {
    root = node;
}

TreeNode* Tree::getRoot() const {
    return root;
}

void Tree::setLeftChild(TreeNode* parent, TreeNode* child) {
    if (parent) parent->left = child;
}

void Tree::setRightChild(TreeNode* parent, TreeNode* child) {
    if (parent) parent->right = child;
}

TreeNode* Tree::getLeftChild(TreeNode* parent) const {
    return parent ? parent->left : nullptr;
}

TreeNode* Tree::getRightChild(TreeNode* parent) const {
    return parent ? parent->right : nullptr;
}

void Tree::preorder(TreeNode* node, std::function<void(char)> visit) const {
    if (node) {
        visit(node->label);
        preorder(node->left, visit);
        preorder(node->right, visit);
    }
}

void Tree::preorder(std::function<void(char)> visit) const {
    preorder(root, visit);
}

void Tree::inorder(TreeNode* node, std::function<void(char)> visit) const {
    if (node) {
        inorder(node->left, visit);
        visit(node->label);
        inorder(node->right, visit);
    }
}

void Tree::inorder(std::function<void(char)> visit) const {
    inorder(root, visit);
}

void Tree::postorder(TreeNode* node, std::function<void(char)> visit) const {
    if (node) {
        postorder(node->left, visit);
        postorder(node->right, visit);
        visit(node->label);
    }
}

void Tree::postorder(std::function<void(char)> visit) const {
    postorder(root, visit);
}

void Tree::clear() {
    delete root; // TreeNode сам удалит поддерево в своём деструкторе
    root = nullptr;
}

bool Tree::isEmpty() const {
    return root == nullptr;
}